### ISSUE TITLE
Cache Crawler 404

### ISSUE THREAD

**Issue by mrfrancescomoro** [funkhaus/vuepress#94]
Hi,
I'm having trouble with the instructions on the wiki, I don't know if I'm doing it wrong or there's actually a bug.

After the AboutUs template part of the wiki I get this error:
![image](https://user-images.githubusercontent.com/2639793/38359281-2e417f30-38c8-11e8-9232-40f62779d008.png)

Thanks.

**Comment by SaFrMo** [comment]
Hi! You might be getting this if there's no actual page at `localhost/about` - we need to figure out a way to handle 404s in the cache-crawler better 😄 

Try to navigate to `localhost/about` in your browser, and if it's not there, create and publish page called About in your WordPress backend. Let us know if that doesn't fix the issue and we'll dig deeper!

**Comment by mrfrancescomoro** [comment]
Yeah, there's no localhost/about page because the root of the website is localhost/amt/ so the about page should work on localhost/amt/about
localhost is my xampp http folder where I have tons of other websites.

the page "About" already exists in wp. I've created the AboutUs.vue file and addedn to the custom-vuepress-templates.php file

the page is "working" because i get the post.content but nothing else. (such as the router-link...i'm following the wiki).

I hope I made myself clear enough with what's the problem here 😅

**Comment by SaFrMo** [comment]
Definitely, this all makes sense!

It looks like the issue is the extra level between `localhost` and `about` - since the cache-crawler uses relative paths, it's treating `/amt/` like it should be ignored, when it's in fact the root of the site.

This also explains why router-links don't work. If you click on one for "About," you'll be taken to `localhost/about` instead of `localhost/amt/about`, since router-links are all relative as well.

@johndigital, one fix I can think of is a field on the backend called "Site Root" - it can default to `/` but allows you to enter any string, and all `router-link`s and `cache-crawler` requests can prepend the site root to their relative paths. I've used the same XAMPP setup on Windows, so this'll be a common enough problem that I think it's worth fixing in the boilerplate. Any alternate fixes you can think of?

@mrfrancescomoro, there's no fix that'll cover all situations at the moment - sorry about that, this is a sneaky bug! As a stopgap, you can do two things:

1. Manually add `/amt` to the start of any router-link `to` props:
    ```
    <router-link to="/about">About</router-link>
    ```

    becomes

    ```
    <router-link to="/amt/about">About</router-link>
    ```

1. Change

    ```
    cache[path] = await fetch(`${path}?contentType=json`).then(r => r.json())
    ```

    to

    ```
    let localPath = '/amt' + path
    cache[localPath] = await fetch(`${localPath}?contentType=json`).then(r => r.json())
    ```

    in `src/utils/cache-crawler.js`.

Thanks again for pointing this out, this is something we never would've caught!

**Comment by mrfrancescomoro** [comment]
@SaFrMo thanks I'll try right away.
I know, of course, that there's no solution for all cases. "Happy" to find new bugs for you to solve 😂 
I'll let you know if it works

**Comment by johndigital** [comment]
Hmm, I think there's actually a simpler fix here. In order to generate those relative URLs we are just taking the permalink of the page and stripping out the Wordpress "[siteurl](https://codex.wordpress.org/Function_Reference/site_url)". The problem is that we need to be stripping out only the domain rather than the whole siteurl because in @mrfrancescomoro's case he has siteurl set to include the `/amt` which of course is a perfectly valid way to configure Wordpress just not something we took into account.

So what we need to do is take [this function](https://github.com/funkhaus/Rest-Easy/blob/b442ef1bd8dcd16d5be6730058b90cdfa1014dc6/utils.php#L6) and modify it so that only the domain is removed. If the siteurl is set to `http://example.com/some` and you feed that function the URL `http://example.com/some/url` it should return `/some/url`

Not sure what the most Wordpress friendly was is to get only the domain, but that should fix it.

**Comment by johndigital** [comment]
Found our answer! https://codex.wordpress.org/Function_Reference/wp_make_link_relative

**Comment by johndigital** [comment]
This has now been fixed in 5c6ce3ea650db916b3d7e258f21b6f716b1ada3a, but requires Rest Easy to be upgraded to the latest commit: https://github.com/funkhaus/Rest-Easy/commit/655b7ccb3bb7e9987ffdd99a3b6ec9c24942da6b

@mrfrancescomoro I tested locally with the URL configuration you described and it should work now. You can leave Vuepress as it is and just replace your [Rest Easy](https://github.com/funkhaus/Rest-Easy) install with the latest code and it should just work!

Please re-open this issue if that doesn't solve this problem for you.
